src/marqo/core/index_management/index_management.py

```diff
@@ -86,15 +86,24 @@ def bootstrap_vespa(self) -> bool:
         Returns:
             True if Vespa was bootstrapped, False if it was already up-to-date
         """
-        with self._vespa_deployment_lock():
-            vespa_app = self._get_vespa_application(check_configured=False, need_binary_file_support=True)
 
-            to_version = version.get_version()
-            from_version = vespa_app.get_marqo_config().version if vespa_app.is_configured else None
+        # We skip the Vespa convergence check here so that Marqo instance can be bootstrapped even when Vespa is
+        # not converged. This will reduce the risk of having downtime when the Marqo instance crashes while waiting
+        # for Vespa app to converge.
+        to_version = version.get_version()
+        vespa_app_for_version_check = self._get_vespa_application(check_configured=False, need_binary_file_support=True,
+                                                                  check_for_application_convergence=False)
+        from_version = vespa_app_for_version_check.get_marqo_config().version \
+            if vespa_app_for_version_check.is_configured else None
+
+        if from_version and semver.VersionInfo.parse(from_version) >= semver.VersionInfo.parse(to_version):
+            # skip bootstrapping if already bootstrapped to this version or later
+            return False
 
-            if from_version and semver.VersionInfo.parse(from_version) >= semver.VersionInfo.parse(to_version):
-                # skip bootstrapping if already bootstrapped to this version or later
-                return False
+        with self._vespa_deployment_lock():
+            # Initialise another session based on the current active session
+            vespa_app = self._get_vespa_application(check_configured=False, need_binary_file_support=True,
+                                                    check_for_application_convergence=False)
```

Comment:
I'm curious to know if we need to call _get_vespa_application() twice? Once here and previously on L#94

Reply:
Hello @adityabharadwaj198 , good question. The reason we call it inside the distributed locking block is that the application package might be changed by other operations when we wait to acquire the lock. If we use the previous vespa session, it is more likely to fail the deployment due to race condition.

Reply:
@papa99do perhaps clarify this in code so others don't change it later